zproject/backends.py

```diff
@@ -866,6 +884,62 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserException as e:
             raise ZulipLDAPException(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if not settings.LDAP_SYNC_GROUPS:
+            return
+
+        if user_profile.realm.id != settings.LDAP_GROUP_SYNC_REALM_ID:
+            return
+
+        try:
+            ldap_logger.debug("found groups for user: %s", str(user_profile))
+            ldap_group_name_set = set(ldap_user.group_names)
+
+            member_group_names = [
+                membership.user_group.name
+                for membership in UserGroupMembership.objects.filter(user_profile=user_profile)
+            ]
+
+            ldap_logger.debug("user groups: %s", repr(member_group_names))
+
+            for group_name in ldap_group_name_set:
```

Comment:
It might be a lot more efficient to do something like `for group in UserGroup.objects.filter(realm=user_profile.realm, name__in=ldap_group_name_set.intersect(set(settings.LDAP_GROUP_SYNC_LIST))` to avoid doing those `.get()`s below in a loop?

Reply:
see https://github.com/zulip/zulip/pull/23039/commits/a159320aa447baea2d27953f7e3462c59cdf5fad